p2p/node/api.go

```diff
@@ -53,9 +53,19 @@ func (p *P2PNode) Subscribe(location common.Location, datatype interface{}) erro
 	}
 
 	go func() {
-		err := p.peerManager.Provide(p.ctx, location, datatype)
-		if err != nil {
-			log.Global.Errorf("error providing topic %s in %s: %s", reflect.TypeOf(datatype), location.Name(), err.Error())
+		ticker := time.NewTicker(time.Second)
+		timeout := time.NewTicker(60 * time.Second)
+		for {
+			select {
+			case <-ticker.C:
+				if err := p.peerManager.Provide(p.ctx, location, datatype); err == nil {
+					log.Global.Infof("providing topic %s in %s", reflect.TypeOf(datatype), location.Name())
+					return
+				}
+			case <-timeout.C:
```

Comment:
The way its written there's a potential issue here if the timeout signal occurs while provide is being processed (synchronously), which could happen repeatedly depending on the length of provide.

Do you know how long provide takes to execute?

Reply:
what is the issue?

Unless I am misunderstanding, if a timeout event occurs while I am actively processing a call to provide(), one of two things happen:
1) provide is successful -> go routine exits without handling timeout event :heavy_check_mark: 
2) provide is unsuccessful -> go routine services the select block again, at which point it sees the pending timeout event

what am I missing?

Reply:
My understanding is in golang the event gets thrown out if the channel is not immediately available for processing

Reply:
I think you're right honestly, ChatGPT might be hallucinating again

ChatGPT
No, the ticker will not queue up if it occurs while Provide is being processed. In Go, select statements do not queue channel operations.

Reply:
That's not how golang channels work. For an unbuffered channel, the sender will block until the receiver is ready. For a buffered channel, the sender can asynchronously proceed if there is room in the buffer. In either case, the receiver will see the value available in the channel on the next iteration.